File: fota-gateway-tcp-server/src/main/java/com/yef/service/PacketSender.java
package com.yef.service;

import com.yef.protocol.FotaMessage;
import com.yef.session.DeviceSession;
import com.yef.session.SessionManager;
import org.springframework.stereotype.Service;

/**
 * @description: 统一发上行消息
 * @author: 叶丰
 * @date: 2026/4/17 09:35
 */
@Service
public class PacketSender {

    private final SessionManager sessionManager;

    public PacketSender(SessionManager sessionManager) {
        this.sessionManager = sessionManager;
    }

    public boolean sendToDevice(String imei, FotaMessage message) {
        DeviceSession session = sessionManager.getByImei(imei);
        if (session == null || session.getChannel() == null || !session.getChannel().isActive()) {
            System.out.println("[PacketSender] device offline, imei=" + imei);
            return false;
        }
        session.getChannel().writeAndFlush(message);
        return true;
    }
}

File: golang-gateway-tcp-server/internal/service/packet_sender.go
package service

import (
	"log"

	"golang-gateway-tcp-server/internal/protocol"
	"golang-gateway-tcp-server/internal/session"
)

type PacketSender struct {
	sessions *session.Manager
}

func NewPacketSender(sessions *session.Manager) *PacketSender {
	return &PacketSender{sessions: sessions}
}

func (s *PacketSender) SendToDevice(imei string, msg protocol.Message) bool {
	deviceSession := s.sessions.GetByIMEI(imei)
	if deviceSession == nil || deviceSession.Conn == nil {
		log.Printf("[PacketSender] device offline, imei=%s", imei)
		return false
	}
	frame, err := protocol.EncodeMessage(msg)
	if err != nil {
		log.Printf("encode outbound message failed, imei=%s err=%v", imei, err)
		return false
	}
	if _, err := deviceSession.Conn.Write(frame); err != nil {
		log.Printf("write outbound message failed, imei=%s err=%v", imei, err)
		return false
	}
	return true
}
